cache/network/api.go

```diff
@@ -143,16 +155,80 @@ func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, he
 	c.logger.Debugf("Response dump: %s", string(dump))
 
 	if resp.StatusCode != http.StatusOK {
-		return unwrapError(resp)
+		return "", unwrapError(resp)
 	}
 
-	return nil
+	etag := resp.Header.Get("ETag")
+
+	return etag, nil
 }
 
-func (c apiClient) acknowledgeUpload(uploadID string) (acknowledgeResponse, error) {
-	url := fmt.Sprintf("%s/upload/%s/acknowledge", c.baseURL, uploadID)
+func (c apiClient) uploadArchive(archivePath string, chunkSize, chunkCount, lastChunkSize int, uploadURLs []uploadURL) ([]string, error) {
+	file, err := os.Open(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("open file: %s", err)
+	}
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			c.logger.Errorf("failed to close file: %s", err)
+		}
+	}(file)
+
+	etags := make([]string, 0, chunkCount)
+
+	if chunkCount == 1 {
+		fileInfo, err := os.Stat(archivePath)
+		if err != nil {
+			return nil, fmt.Errorf("stat file: %s", err)
+		}
+		c.logger.Debugf("Uploading single chunk (non multipart upload)", uploadURLs[0].URL)
+
+		_, err = c.uploadArchiveChunk(uploadURLs[0], file, fileInfo.Size())
+		if err != nil {
+			return nil, fmt.Errorf("upload single chunk (non multipart): %s", err)
+		}
+		return nil, nil
+	}
+
+	c.logger.Debugf("Uploading %d chunks, %dB each", chunkCount, chunkSize)
+
+	for i := 0; i < chunkCount; i++ {
+		chunkData, err := io.ReadAll(io.NewSectionReader(file, int64(i)*int64(chunkSize), int64(chunkSize)))
+		if err != nil {
+			return nil, fmt.Errorf("read chunk: %s", err)
+		}
+
+		if i < chunkCount-1 && len(chunkData) != chunkSize {
+			c.logger.Warnf("chunk size mismatch, expected %d, got %d", chunkSize, len(chunkData))
+		}
+		if i == chunkCount-1 && len(chunkData) != lastChunkSize {
+			c.logger.Warnf("last chunk size mismatch, expected %d, got %d", lastChunkSize, len(chunkData))
+		}
+
+		c.logger.Debugf("Uploading chunk %d to %s", i, uploadURLs[i].URL)
+		etag, err := c.uploadArchiveChunk(uploadURLs[i], chunkData, int64(len(chunkData)))
```

Comment:
do we plan to do it in parallel in future iterations?

Reply:
Yeah it would be a great improvement later